Sources/NIOCore/ChannelPipeline.swift

```diff
@@ -2089,3 +2089,171 @@ extension ChannelPipeline: CustomDebugStringConvertible {
+    ///            `NIOOutboundByteBufferingChannelHandler`, the future will contain`nil`.
+    public func outboundBufferedBytes(in context: ChannelHandlerContext) -> EventLoopFuture<Int?> {
+        let future: EventLoopFuture<Int?>
+
+        if self.eventLoop.inEventLoop {
+            future = self.eventLoop.makeSucceededFuture(countBufferedBytes(context: context, direction: .outbound))
+        } else {
+            future = self.eventLoop.submit {
+                self.countBufferedBytes(context: context, direction: .outbound)
+            }
+        }
+
+        return future
+    }
+
+    /// Retrieve the total number of bytes buffered for inbound.
+    public func inboundBufferedBytes() -> EventLoopFuture<Int> {
+        let future: EventLoopFuture<Int>
+
+        if self.eventLoop.inEventLoop {
+            future = self.eventLoop.makeSucceededFuture(countAllBufferedBytes(direction: .inbound))
+        } else {
+            future = self.eventLoop.submit {
+                self.countAllBufferedBytes(direction: .inbound)
+            }
+        }
+
+        return future
+    }
+
+    /// Retrieve the number of inbound bytes buffered in the `ChannelHandler` associated with the given`ChannelHandlerContext`.
+    ///
+    /// - Parameters:
+    ///     - in: the `ChannelHandlerContext` from which the inbound buffered bytes of the `ChannelHandler` will be retrieved.
+    ///
+    /// - Returns: The `EventLoopFuture` which will be notified once the number of inbound bytes buffered in the `ChannelHandler`
+    ///            referenced by the `ChannelHandlerContext` parameter `in` is collected.
+    ///            If the `ChannelHandler` in the `ChannelHandlerContext` does not conform to
+    ///            `NIOInboundByteBufferingChannelHandler`, the future will contain `nil`.
+    public func inboundBufferedBytes(in context: ChannelHandlerContext) -> EventLoopFuture<Int?> {
+        let future: EventLoopFuture<Int?>
+
+        if self.eventLoop.inEventLoop {
+            future = self.eventLoop.makeSucceededFuture(countBufferedBytes(context: context, direction: .inbound))
+        } else {
+            future = self.eventLoop.submit {
+                self.countBufferedBytes(context: context, direction: .inbound)
+            }
+        }
+
+        return future
+    }
+
+    private func countBufferedBytes(context: ChannelHandlerContext, direction: BufferingDirection) -> Int? {
+        switch direction {
+        case .inbound:
+            guard let handler = context.handler as? NIOInboundByteBufferingChannelHandler else {
+                return nil
+            }
+            return handler.inboundBufferedBytes
+        case .outbound:
+            guard let handler = context.handler as? NIOOutboundByteBufferingChannelHandler else {
+                return nil
+            }
+            return handler.outboundBufferedBytes
+        }
+
+    }
+
+    private func countAllBufferedBytes(direction: BufferingDirection) -> Int {
```

Comment:
Can we add an `assertInEventLoop` here?

Reply:
fixed in https://github.com/apple/swift-nio/pull/2918/commits/84b9fd438c4f93fb16fcdc2b85a4af6eb5aae48e